xapian-letor/feature/colltfcolllenfeature.cc

```diff
@@ -38,37 +38,74 @@ CollTfCollLenFeature::name() const
 vector<double>
 CollTfCollLenFeature::get_values() const
 {
-    LOGCALL(API, std::vector<double>, "CollTfCollLenFeature::get_values", NO_ARGS);
+    LOGCALL(API, vector<double>, "CollTfCollLenFeature::get_values", NO_ARGS);
 
     vector<double> values;
     double value = 0;
+    double coll_len = 0;
+    map<string, Xapian::termcount>::const_iterator coll_len_iterator =
+	    collection_length.find("title");
+    if (coll_len_iterator != collection_length.end())
+	coll_len = (double)coll_len_iterator->second;
+    else
+	coll_len = 0;
 
     for (Xapian::TermIterator qt = feature_query.get_unique_terms_begin();
 	 qt != feature_query.get_terms_end(); ++qt) {
-	if ((*qt).substr(0, 1) == "S" || (*qt).substr(1, 1) == "S")
-	    value += log10(1 + ((double)collection_length.find("title")->second
-				/ (double)(1 + collection_termfreq.find(*qt)->second)));
+	if ((*qt).substr(0, 1) == "S" || (*qt).substr(1, 1) == "S") {
```

Comment:
The `substr(0, 1) == "S"` check I understand - that's a hard-coded check for this being the title field (though it'd be better written as `startswith(*qt, 'S')` - `startswith()` is a utility function in `"stringutils.h"` which can check for a string starting with a string or a character without creating the temporary string object which the current code does).

But `substr(1, 1) == "S"` is wrong (sorry, I missed this amongst all the stylistic issues).  That's checking if the second character of the term is `S`, but it's really wanting to check for terms with prefix `ZS` (stemmed terms from the title), so it ought to check `startswith(*qt, "ZS")`.  As-is, it will match those but also unwanted terms like `XSUMMARYfoo` (term "foo" with user prefix "XSUMMARY") and `KSI units` (keyword "SI units").

Given how many times this check is used, let's turn it into an inline helper:

```
inline bool
is_title_term(const string& term)
{
    return startswith(term, 'S') || startswith(term, "ZS");
}
```

That'll improve both readability and maintainability.